File: trunk/ShadowFrameworkJS/src/shadow/system/data/SFGenericDatasetFactory.js

function SFGenericDatasetFactory(){
	this.datasets=new Array();
	this.assetsList=new Array();
}

SFGenericDatasetFactory.prototype["addSFDataset"]=function(sfDataset){
	var typeName=getType(sfDataset);
	this.datasets[typeName]=sfDataset;
	this.assetsList.push(sfDataset);
};

SFGenericDatasetFactory.prototype["readDataset"]=function(stream){
	var index=stream.readInt();
	var dataset = this.assetsList[index];
	dataset = dataset.generateNewDatasetInstance();
	//alert("readDataset "+(dataset.generateNewDatasetInstance)+" "+index);
	dataset.getSFDataObject().readFromStream(stream);
	//alert("It's been read!! "+index);
	return dataset;
};	

SFGenericDatasetFactory.prototype["writeDataset"]=function(stream,dataset){
			stream.writeString(dataset.getType());
			dataset.getSFDataObject().writeOnStream(stream);
		};
	
SFGenericDatasetFactory.prototype["createDataset"]=function(typeName){
			//TODO : there is an error here... this is not used...
			var asset=this.datasets[typeName];
			if(asset != undefined){
				return asset.generateNewDatasetInstance();
			}
			throw "Cannot instantiate "+asset;
		};

File: SFDataApplicationTestC++/shadow/system/data/SFGenericDatasetFactory.cpp
//
//  SFGenericDatasetFactory.cpp
//  
//
//  Created by Alessandro Martinelli on 19/10/12.
//  Copyright (c) 2012 Alessandro Martinelli. All rights reserved.
//

#include "SFGenericDatasetFactory.h"


namespace sf{

SFGenericDatasetFactory::SFGenericDatasetFactory() {
}

SFGenericDatasetFactory::~SFGenericDatasetFactory(){

    for (unsigned int i=0; i<assetsList.size(); i++) {
        delete assetsList[i];
    }
}

template <class T>
void SFGenericDatasetFactory::addSFDataset(SFDataAsset<T>* sfasset) {
    assetsList.push_back(sfasset);
}


SFDataset* SFGenericDatasetFactory::readDataset(SFInputStream* stream) {
    int index=stream->readInt();
    SFDataset* dataset=assetsList.at(index);
    dataset = dataset->generateNewDatasetInstance();

    dataset->getSFDataObject()->readFromStream(stream);
    return dataset;
}

template <class T>
	SFDataAsset<T>* SFGenericDatasetFactory::createDataset(int index){

		SFDataAsset<T>* asset=assetsList[index];

		return asset->generateNewDatasetInstance();
	}

}
